chia/wallet/trade_manager.py

```diff
@@ -178,516 +136,338 @@ async def get_locked_coins(self, wallet_id: int = None) -> Dict[bytes32, WalletC
 
         return result
 
     async def get_all_trades(self):
         all: List[TradeRecord] = await self.trade_store.get_all_trades()
         return all
 
-    async def get_trade_by_id(self, trade_id: bytes) -> Optional[TradeRecord]:
+    async def get_trade_by_id(self, trade_id: bytes32) -> Optional[TradeRecord]:
         record = await self.trade_store.get_trade_record(trade_id)
         return record
 
-    async def get_locked_coins_in_spend_bundle(self, bundle: SpendBundle) -> Dict[bytes32, WalletCoinRecord]:
-        """Returns a list of coin records that are used in this SpendBundle"""
-        result = {}
-        removals = bundle.removals()
-        for coin in removals:
-            coin_record = await self.wallet_state_manager.coin_store.get_coin_record(coin.name())
-            if coin_record is None:
-                continue
-            result[coin_record.name()] = coin_record
-        return result
-
     async def cancel_pending_offer(self, trade_id: bytes32):
-        await self.trade_store.set_status(trade_id, TradeStatus.CANCELED, False)
+        await self.trade_store.set_status(trade_id, TradeStatus.CANCELLED, False)
 
-    async def cancel_pending_offer_safely(self, trade_id: bytes32):
+    async def cancel_pending_offer_safely(
+        self, trade_id: bytes32, fee: uint64 = uint64(0)
+    ) -> Optional[List[TransactionRecord]]:
         """This will create a transaction that includes coins that were offered"""
         self.log.info(f"Secure-Cancel pending offer with id trade_id {trade_id.hex()}")
         trade = await self.trade_store.get_trade_record(trade_id)
         if trade is None:
             return None
 
-        all_coins = trade.removals
-
-        for coin in all_coins:
+        all_txs: List[TransactionRecord] = []
+        fee_to_pay: uint64 = fee
+        for coin in Offer.from_bytes(trade.offer).get_primary_coins():
             wallet = await self.wallet_state_manager.get_wallet_for_coin(coin.name())
 
             if wallet is None:
                 continue
             new_ph = await wallet.get_new_puzzlehash()
-            if wallet.type() == WalletType.COLOURED_COIN.value:
+            # This should probably not switch on whether or not we're spending a CAT but it has to for now
+            if wallet.type() == WalletType.COLOURED_COIN:
                 txs = await wallet.generate_signed_transaction(
-                    [coin.amount], [new_ph], 0, coins={coin}, ignore_max_send_amount=True
+                    [coin.amount], [new_ph], fee=fee_to_pay, coins={coin}, ignore_max_send_amount=True
                 )
-                tx = txs[0]
+                all_txs.extend(txs)
             else:
                 tx = await wallet.generate_signed_transaction(
-                    coin.amount, new_ph, 0, coins={coin}, ignore_max_send_amount=True
+                    uint64(coin.amount - fee_to_pay), new_ph, fee=fee_to_pay, coins={coin}, ignore_max_send_amount=True
```

Comment:
Is there a possible issue if fee_to_pay is > than coin.amount? I tried canceling an offer using a fee > the offered amount and didn't observe any issues, but perhaps that experiment didn't exercise this code path.

Reply:
Oh hmm this is an interesting potential bug in the software.  We have to specify the coin to use because we must spend it but if the fee is higher, we need logic to select more coins.  Gross, but I'll take a look at it.